File: Solutions/LeetCode/0002-AddTwoNumbers/Solution.cs
using System;
using System.Collections.Generic;

public class ListNode {
    public int val;
    public ListNode next;
    public ListNode(int val=0, ListNode next=null) {
        this.val = val;
        this.next = next;
    }
}

public class Solution {
    public ListNode AddTwoNumbers(ListNode l1, ListNode l2) {
        ListNode dummy = new ListNode(0);
        dummy.next = l1;
        ListNode curr = dummy;
        int carry = 0;
        
        while (l1 != null || l2 != null || carry != 0) {
            int val = carry;
            if (l1 != null) {
                val += l1.val;
            }
            if (l2 != null) {
                val += l2.val;
            }
            
            carry = val / 10;
            val = val % 10;
            
            if (l1 != null) {
                l1.val = val;
                curr = l1;
                l1 = l1.next;
            } else if (l2 != null) {
                l2.val = val;
                curr.next = l2;
                curr = l2;
                l2 = l2.next;
            } else {
                curr.next = new ListNode(val);
                curr = curr.next;
            }
            
            if (l2 != null && l1 == null) {
                curr.next = l2;
                l1 = l2;
                l2 = null;
            }
        }
        return dummy.next;
    }

    public static ListNode ToListNode(int[] arr) {
        ListNode dummy = new ListNode();
        ListNode curr = dummy;
        foreach (int val in arr) {
            curr.next = new ListNode(val);
            curr = curr.next;
        }
        return dummy.next;
    }

    public static List<int> ToList(ListNode node) {
        List<int> res = new List<int>();
        while (node != null) {
            res.Add(node.val);
            node = node.next;
        }
        return res;
    }

    public static void Main(string[] args) {
        Solution sol = new Solution();
        var testCases = new (int[] l1, int[] l2, int[] expected)[] {
            (new int[] { 2, 4, 3 }, new int[] { 5, 6, 4 }, new int[] { 7, 0, 8 }),
            (new int[] { 0 }, new int[] { 0 }, new int[] { 0 }),
            (new int[] { 9, 9, 9, 9, 9, 9, 9 }, new int[] { 9, 9, 9, 9 }, new int[] { 8, 9, 9, 9, 0, 0, 0, 1 })
        };

        for (int i = 0; i < testCases.Length; i++) {
            var tc = testCases[i];
            ListNode l1 = ToListNode(tc.l1);
            ListNode l2 = ToListNode(tc.l2);
            ListNode resNode = sol.AddTwoNumbers(l1, l2);
            List<int> result = ToList(resNode);
            bool pass = result.Count == tc.expected.Length;
            if (pass) {
                for (int j = 0; j < result.Count; j++) {
                    if (result[j] != tc.expected[j]) {
                        pass = false;
                        break;
                    }
                }
            }
            Console.WriteLine($"Test {i + 1}: l1=[{string.Join(", ", tc.l1)}], l2=[{string.Join(", ", tc.l2)}] -> Result=[{string.Join(", ", result)}] | Expected=[{string.Join(", ", tc.expected)}] | {(pass ? "PASS" : "FAIL")}");
        }
    }
}

File: Solutions/LeetCode/0002-AddTwoNumbers/Solution.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(0)
        dummy.next = l1
        curr = dummy
        carry = 0
        
        while l1 or l2 or carry:
            val = carry
            if l1:
                val += l1.val
            if l2:
                val += l2.val
                
            carry, val = divmod(val, 10)
            
            if l1:
                l1.val = val
                curr = l1
                l1 = l1.next
                if l2:
                    l2 = l2.next
            elif l2:
                l2.val = val
                curr.next = l2
                curr = l2
                l2 = l2.next
            else:
                curr.next = ListNode(val)
                curr = curr.next
                
            if l2 and not l1:
                curr.next = l2
                l1 = l2
                l2 = None
                
        return dummy.next

def to_list_node(lst):
    dummy = ListNode()
    curr = dummy
    for val in lst:
        curr.next = ListNode(val)
        curr = curr.next
    return dummy.next

def to_array(node):
    res = []
    while node:
        res.append(node.val)
        node = node.next
    return res

if __name__ == '__main__':
    sol = Solution()
    test_cases = [
        ([2, 4, 3], [5, 6, 4], [7, 0, 8]),
        ([0], [0], [0]),
        ([9, 9, 9, 9, 9, 9, 9], [9, 9, 9, 9], [8, 9, 9, 9, 0, 0, 0, 1])
    ]
    for idx, (l1_arr, l2_arr, expected) in enumerate(test_cases, 1):
        l1 = to_list_node(l1_arr)
        l2 = to_list_node(l2_arr)
        res_node = sol.addTwoNumbers(l1, l2)
        result = to_array(res_node)
        print(f"Test {idx}: l1={l1_arr}, l2={l2_arr} -> Result={result} | Expected={expected} | {'PASS' if result == expected else 'FAIL'}")